File: backend/src/main/java/com/safesite/model/Report.java
package com.safesite.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "reports")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Report {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    
    @Column(nullable = false)
    private String title;
    
    @Column(columnDefinition = "TEXT")
    private String content;
    
    @Column(name = "report_type")
    @Enumerated(EnumType.STRING)
    private ReportType reportType;
    
    @Column(name = "report_date")
    private LocalDate reportDate;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "project_id")
    private Project project;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "site_id")
    private Site site;
    
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by")
    private User createdBy;
    
    @Column(name = "file_url")
    private String fileUrl; // PDF file URL
    
    @Column(name = "created_at")
    private LocalDateTime createdAt;
    
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;
    
    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        if (reportDate == null) {
            reportDate = LocalDate.now();
        }
    }
    
    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
    
    public enum ReportType {
        DAILY, WEEKLY, MONTHLY, INCIDENT, SAFETY_AUDIT
    }
}

File: frontend/lib/core/models/report.dart
import 'project.dart';
import 'site.dart';
import 'user.dart';

enum ReportType {
  daily,
  weekly,
  monthly,
  incident,
  safetyAudit;

  static ReportType fromString(String value) {
    switch (value.toUpperCase()) {
      case 'DAILY':
        return ReportType.daily;
      case 'WEEKLY':
        return ReportType.weekly;
      case 'MONTHLY':
        return ReportType.monthly;
      case 'INCIDENT':
        return ReportType.incident;
      case 'SAFETY_AUDIT':
        return ReportType.safetyAudit;
      default:
        return ReportType.daily;
    }
  }

  String toApiString() {
    switch (this) {
      case ReportType.daily:
        return 'DAILY';
      case ReportType.weekly:
        return 'WEEKLY';
      case ReportType.monthly:
        return 'MONTHLY';
      case ReportType.incident:
        return 'INCIDENT';
      case ReportType.safetyAudit:
        return 'SAFETY_AUDIT';
    }
  }
}

class Report {
  final int? id;
  final String title;
  final String? content;
  final ReportType reportType;
  final DateTime? reportDate;
  final int? projectId;
  final Project? project;
  final int? siteId;
  final Site? site;
  final int? createdById;
  final User? createdBy;
  final String? fileUrl;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  Report({
    this.id,
    required this.title,
    this.content,
    required this.reportType,
    this.reportDate,
    this.projectId,
    this.project,
    this.siteId,
    this.site,
    this.createdById,
    this.createdBy,
    this.fileUrl,
    this.createdAt,
    this.updatedAt,
  });

  factory Report.fromJson(Map<String, dynamic> json) {
    return Report(
      id: json['id'],
      title: json['title'] ?? '',
      content: json['content'],
      reportType: ReportType.fromString(json['reportType'] ?? 'DAILY'),
      reportDate: json['reportDate'] != null
          ? DateTime.parse(json['reportDate'])
          : null,
      projectId:
          json['project'] != null ? json['project']['id'] : json['projectId'],
      project: json['project'] != null ? Project.fromJson(json['project']) : null,
      siteId: json['site'] != null ? json['site']['id'] : json['siteId'],
      site: json['site'] != null ? Site.fromJson(json['site']) : null,
      createdById:
          json['createdBy'] != null ? json['createdBy']['id'] : json['createdById'],
      createdBy:
          json['createdBy'] != null ? User.fromJson(json['createdBy']) : null,
      fileUrl: json['fileUrl'],
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'content': content,
      'reportType': reportType.toApiString(),
      'reportDate': reportDate?.toIso8601String().split('T')[0],
      'project': projectId != null ? {'id': projectId} : project?.toJson(),
      'site': siteId != null ? {'id': siteId} : site?.toJson(),
      'fileUrl': fileUrl,
    };
  }
}

